gadget/update.go

```diff
@@ -326,6 +327,111 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
 	return nil
 }
 
+// DiskTraitsFromDeviceAndValidate takes a laid out gadget volume and an
+// expected disk device path and confirms that they are compatible, and then
+// builds up the disk volume traits for that device. If the laid out volume is
+// not compatible with the disk structure for the specified device an error is
+// returned.
+func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string) (res DiskVolumeDeviceTraits, err error) {
+	vol := expLayout.Volume
+
+	// get the disk layout for this device
+	diskLayout, err := OnDiskVolumeFromDevice(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot read %v partitions for candidate volume %s: %v", dev, vol.Name, err)
+	}
+
+	// ensure that the on disk volume and the laid out volume are actually
+	// compatible
+	opts := &EnsureLayoutCompatibilityOptions{
+		// at this point all partitions should be created
+		AssumeCreatablePartitionsCreated: true,
+	}
+	if err := EnsureLayoutCompatibility(expLayout, diskLayout, opts); err != nil {
+		return res, fmt.Errorf("volume %s is not compatible with disk %s: %v", vol.Name, dev, err)
+	}
+
+	// also get a Disk{} interface for this device
+	disk, err := disks.DiskFromDeviceName(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot get disk for device %s: %v", dev, err)
+	}
+
+	mappedStructures := make([]DiskStructureDeviceTraits, 0, len(diskLayout.Structure))
```

Comment:
nitpick: I'd move this down below, right before the `for` loop where it's used.

Reply:
done